### PR TITLE
Add restored resource list in the output of restore describe command

### LINKED ISSUE #5199: Add restored resource list with a restore job
**Describe the problem/challenge you have**

Now velero can overwrite part of existing resources during a restore by `ExistingResourcePolicy`.  
This is very useful for in-place restore. 

However, there is lack of report for user to inspect the result of restore, especially when restore fails.  

Ideally we could have a restore resource list, like current backup details, to show which resource is created/overwritten/skipped during this restore job.  So user could understand if any resource in her/his cluster is modified. 

**Describe the solution you'd like**

Similar to `BackupResourceList` mechanism,  velero could record the restore action/result on each resource and upload the record json file to S3. The corresponding cli or download request can download it to users.

**Anything else you would like to add:**

NA

**Environment:**

- Velero version (use `velero version`):
- Kubernetes version (use `kubectl version`):
- Kubernetes installer & version:
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "The project would be better with this feature added"
- :-1: for "This feature will not enhance the project in a meaningful way"

### PR DESCRIPTION
**Pull request by ywk253100**
Add restored resource list in the output of restore describe command

Fixes #5199

Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #(issue)

# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Comment by ywk253100** [comment]
The restored resource list is as follows:
```
velero restore describe r04 --details
Name:         r04
Namespace:    velero
Labels:       <none>
Annotations:  <none>

Phase:                       Completed
Total items to be restored:  12
Items restored:              12

Started:    2023-02-16 16:03:17 +0800 CST
Completed:  2023-02-16 16:03:57 +0800 CST

Backup:  restore-list-05

Namespaces:
  Included:  all namespaces found in the backup
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io, backups.velero.io, restores.velero.io, resticrepositories.velero.io, csinodes.storage.k8s.io, volumeattachments.storage.k8s.io, backuprepositories.velero.io
  Cluster-scoped:  auto

Namespace mappings:  <none>

Label selector:  <none>

Restore PVs:  auto

Existing Resource Policy:   update

Preserve Service NodePorts:  auto

Resource List:
  apiextensions.k8s.io/v1/CustomResourceDefinition:
    - crontabs.stable.example.com(updated)
  stable.example.com/v1/CronTab:
    - test/my-new-cron-object(created)
  v1/ConfigMap:
    - test/kube-root-ca.crt(skipped)
  v1/Namespace:
    - test(created)
  v1/PersistentVolume:
    - pvc-938faceb-6efa-438f-8cc3-3e6e4a9487f8(created)
  v1/PersistentVolumeClaim:
    - test/etcd0-pv-claim(created)
  v1/Pod:
    - test/etcd0(created)
  v1/Secret:
    - test/default-token-t8z5d(skipped)
    - test/test(created)
    - test/test-token-kblwh(skipped)
  v1/ServiceAccount:
    - test/default(skipped)
    - test/test(created)
```

File: pkg/restore/restore.go [hunk 1]
```diff
@@ -979,16 +983,29 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
...
 	}
-	ctx.restoredItems[itemKey] = struct{}{}
+	ctx.restoredItems[itemKey] = ""
+	defer func() {
+		// the action field is set explicitly
+		if action := ctx.restoredItems[itemKey]; len(action) > 0 {
+			return
+		}
+		// no action specified, and no warnings and errors
+		if errs.IsEmpty() && warnings.IsEmpty() {
```
Comment (Lyndon-Li) [inline pkg/restore/restore.go:1002]: In which case will this happen? No errors nor warn, but the result was not updated.
Can we deduce the item is skipped for restore for sure when this happens?

File: pkg/restore/restore.go [hunk 2]
```diff
@@ -979,16 +983,29 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
...
 	}
 	if _, exists := ctx.restoredItems[itemKey]; exists {
 		ctx.log.Infof("Skipping %s because it's already been restored.", resourceID)
 		return warnings, errs
 	}
-	ctx.restoredItems[itemKey] = struct{}{}
+	ctx.restoredItems[itemKey] = ""
+	defer func() {
+		// the action field is set explicitly
+		if action := ctx.restoredItems[itemKey]; len(action) > 0 {
```
Comment (Lyndon-Li) [inline pkg/restore/restore.go:998]: Looks like this is the only place to check the value of `ctx.restoredItems`, and it checks if the value is empty or not. Therefore, do we need to define these detailed strings as below?
```
	itemRestoreResultCreated = "created"
	itemRestoreResultUpdated = "updated"
	itemRestoreResultFailed  = "failed"
	itemRestoreResultSkipped = "skipped"
```

File: pkg/restore/restore.go (hunk 1 above)
Comment (ywk253100) [reply pkg/restore/restore.go:1002]: There are several cases in which the resources are skipped, e.g.
- [The resources are excluded by users](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L912-L964)
- [The resource is completed](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L975)
- [Mirror pod](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L996)
- [PV that backed up by PVB](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L1089)
- [PV that has delete reclaim policy](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L1097)
- [Excluded by plugin](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L1151)
- [The resource already exists and is the same as with backup](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L1362)

The codes are in different places and are quite easy to miss out. So here I just specify the action for creation and update, and set the action as skip for others that have no errors and warnings.

**Review by blackpiglet — APPROVED** [review]

**Review by Lyndon-Li — APPROVED** [review]

**Pull request merged**